### PR TITLE
Add `private_ips` attribute to `aws_lb` resource

### PR DESCRIPTION
**Pull request by joshuaspence**
Network load balancers don't have security groups, which makes it difficult to configure security group rules for the backends, which is required for health check to function correctly. According to the [documentation](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-register-targets.html), the way to configure security group rules for health checks involves the following steps:

  1. Open the Amazon EC2 console at https://console.aws.amazon.com/ec2/.
  2. In the navigation pane, choose "Network Interfaces".
  3. In the search field, type the name of your Network Load Balancer. There is one network interface per load balancer subnet.
  4. On the Details tab for each network interface, copy the address from Primary private IPv4 IP.

Obviously, this manual process doesn't play well with Terraform. This issue has also been raised on the [AWS discussion forums](https://forums.aws.amazon.com/thread.jspa?threadID=263245).

This pull request adds a `private_ips` attribute to the `aws_lb` resource to workaround what seems to be an oversight on the part of AWS.

### CODE REVIEW

**Comment by joshuaspence** [comment]
I'm not entirely convinced that this is a good idea, but am looking for feedback.

**Comment by bmathieu33** [comment]
[AWS documentation actually says to do so](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-register-targets.html#target-security-groups):

> If you do not want to grant access to the entire VPC CIDR, you can grant access to the private IP addresses used by the load balancer nodes. There is one IP address per load balancer subnet. 

This is needed to have health check access restriction (for some people its problematic to allow the whole VPC subnet, especially if health check port and application are the same).

File: website/docs/r/lb.html.markdown [hunk 1]
```diff
@@ -54,7 +54,7 @@ Terraform will autogenerate a name beginning with `tf-lb`.
 * `access_logs` - (Optional) An Access Logs block. Access Logs documented below.
 * `subnets` - (Optional) A list of subnet IDs to attach to the LB. Subnets
 cannot be updated for Load Balancers of type `network`. Changing this value
-will for load balancers of type `network` will force a recreation of the resource. 
+will for load balancers of type `network` will force a recreation of the resource.
```
Comment (loivis) [inline website/docs/r/lb.html.markdown]: Do you mind to fix typo here in the mean time? I think there's an extra `will`. It probably should be:

Changing this value for load balancers of type `network` will force a recreation of the resource.

**Comment by ungureanuvladvictor** [comment]
Is there any update on this PR?

Would really love to have access to the private ips generated for the `aws_lb` resource.

**Comment by frittentheke** [comment]
@joshuaspence  Seeing that there already is the array of _LoadBalancerAddresses_  (see: https://forums.aws.amazon.com/thread.jspa?threadID=263245) I wonder if AWS is going to return the private addresses with the ELBv2 query at some point.

**Comment by simonvanderveldt** [comment]
Would really like this functionality because of the use case mentioned in the PRs description.
@joshuaspence have you been using this functionality since you made the PR? How's it working for you?

@frittentheke I wouldn't hold my breath, what AWS puts in their API/responses doesn't really mean a lot. For example: There's been a loadbalancer**s** (plural) key in the ECS service definition as well since the start, hinting at support for multiple loadbalancers, but it still only takes a single loadbalancer.

**Comment by joshuaspence** [comment]
I'm not actually using this functionality at the moment, but when I originally submitted the pull request I did test it out and it appeared to work as I expected.

**Comment by ctreatma** [comment]
Does this PR just need a rebase in order to get merged?  This feature would be very useful & I'm happy to do whatever I can to get it finalized.

**Comment by joshuaspence** [comment]
I'm happy to rebase it, I just haven't bothered as the upstream doesn't seem interested.

**Comment by gnowxilef** [comment]
any plans on rebasing and merging this in? super helpful for a lot of us.

**Comment by ngamber** [comment]
A rebase and merge would be greatly appreciated. Not happy with whitelisting a whole VPC or dealing with the kludgy workarounds.

**Comment by joshuaspence** [comment]
Rebased

**Comment by spa-87** [comment]
@loivis Could your review this PR one more time, please.  It's highly-awaited feature by many people (including me) :). Thanks in advance.

**Comment by oonisim** [comment]
While the feature is not available.
```
data "aws_network_interfaces" "this" {
  filter {
    name = "description"
    values = ["ELB net/${var.alb_nlb_name}/*"]
  }
  filter {
    name = "vpc-id"
    values = ["${local.vpc_chat_service_id}"]
  }
  filter {
    name = "status"
    values = ["in-use"]
  }
  filter {
    name = "attachment.status"
    values = ["attached"]
  }
}

locals {
  nlb_network_interface_ids = "${flatten(["${data.aws_network_interfaces.this.ids}"])}"
}

data "aws_network_interface" "ifs" {
  count = "${length(local.nlb_network_interface_ids)}"
  id = "${local.nlb_network_interface_ids[count.index]}"
}

locals {
  aws_nlb_network_interface_ips = "${flatten([data.aws_network_interface.ifs.*.private_ips])}"
  aws_nlb_network_interface_cidr_blocks = [ for ip in local.aws_nlb_network_interface_ips : "${ip}/32" ]
}
```

**Comment by dlutsch** [comment]
are there any plans to move forward with merging this? This is a feature many of us who use NLBs are waiting for.

**Comment by mdugdale** [comment]
is there any update on this? this feature would be very helpful

**Comment by bazmera** [comment]
Really appreciate if this is merged.It would be very helpful feature.

**Comment by h0nIg** [comment]
@joshuaspence @loivis can you please merge this, this functionality would be very helpful

**Comment by joshuaspence** [comment]
I'm not affiliated with Hashicorp in any way... I don't have permission to merge this.

**Comment by loivis** [comment]
Me neither.

**Comment by sjpalf** [comment]
@maryelizbeth what's needed to get this PR merged (other than another rebase!)? It's been open almost 2 years now, it's a pretty trivial change, and a lot of people would find this really useful!

**Comment by wwboynton** [comment]
Kind of embarrassing how long this PR has been hanging around. @maryelizbeth any chance we can get another set of eyes on this and give @joshuaspence the okay to rebase so we can merge the PR and be done? Unless we're all missing something, this seems like a trivial change with a high positive impact for many of us.

**Comment by h0nIg** [comment]
fixed conflicts: https://github.com/terraform-providers/terraform-provider-aws/pull/11000

**Comment by wwboynton** [comment]
@maryelizbeth can someone from Hashicorp at least comment publicly on why this can't be merged yet?

**Comment by mixedCase** [comment]
A late happy 2 year birthday to this PR! This feature still pretty much mandatory to work with NLBs in many situations and has to be worked around.

**Comment by davidferguson-cr** [comment]
Worth tracking AWS CloudFormation feature request aws-cloudformation/aws-cloudformation-coverage-roadmap#305 for the `AWS::ElasticLoadBalancingV2::LoadBalancer-GetAtt-IpAddressList` capability to see who gets there first...

**Comment by mattrayner** [comment]
Any news on this? I currently need it, or will have to whitelist the whole VPC :/

**Comment by fawaf** [comment]
+1 for merging, but conflicting files again.

**Comment by h0nIg** [comment]
@gdavison @bflad can you please take a look? this is extremly cumbersome, since this PR will bring so much value and nobody is taking a look at it.

**Comment by h0nIg** [comment]
can you please take a look? @bflad @radeksimko @Ninir @ryndaniels @stack72 @nywilken @gdavison @grubernaut @jen20 @catsby @aeschright @paddycarver @appilon @tombuildsstuff @apparentlymart @jbardin @abinashmeher999 @hendrik363

**Comment by binarymist** [comment]
Really need this..

**Comment by strowi** [comment]
Also hoping for this to get merged soon...

**Comment by wwboynton** [comment]
It's one thing for a PR to drag on, it happens...but it's boggling that folks from Hashicorp refuse to even respond to explain why this is the case. Folks like @apparentlymart are usually incredibly kind and pretty upfront about explaining when a PR does not meet the standards or otherwise conflicts with the direction of the project, but this PR has been allowed to go ignored by the Terraform team for multiple years now. It's really, really disappointing behaviour...

**Comment by anthonymade** [comment]
Can somebody please comment on why this PR has not been merged and what would need to happen for this feature to be added?

**Comment by jen20** [comment]
@h0nIg et al: sorry, I am not affiliated with HashiCorp and cannot merge things into the AWS provider!

**Comment by joshuaspence** [comment]
I am also not affiliated with Hashicorp in any way and cannot merge this PR

**Comment by maryelizbeth** [comment]
Hi Y’all, 

First, an apology. This PR got lost in the shuffle and we're sorry for not picking it back up or posting an update in a timely fashion.

We had been punting on this PR because it conflicts with two design principals:
* 	Introduces functionality not present in the ELBv2 API
* 	Introduces cross-service functionality with EC2. We avoid cross-service handling in resources, as it has the potential to introduce confusing permissions and behaviors in constrained environments.

A few months ago the ELBv2 API added the ability to configure the private IPs manually. That functionality is addressed in [#11404](https://github.com/terraform-providers/terraform-provider-aws/pull/11404). 

We’ll work with the author of #11404 to address any changes that may be needed to merge the PR and will merge it once the work is complete. 

As a result, we will close this PR in order to focus conversation on [#11404](https://github.com/terraform-providers/terraform-provider-aws/pull/11404).

If you feel that #11404 does not adequately address your workflow, please open a new GitHub Issue describing the gap. If you have concerns regarding the design of the upstream API, please reach out to AWS.

Once again, we apologize for letting this linger without response and will work to merge [#11404](https://github.com/terraform-providers/terraform-provider-aws/pull/11404) in an upcoming release.

**Pull request closed without merging**